plugin/trino-hive/src/main/java/io/trino/plugin/hive/fs/CachingDirectoryLister.java

```diff
@@ -146,19 +147,23 @@ private static RemoteIterator<TrinoFileStatus> createListingRemoteIterator(Trino
     @Override
     public void invalidate(Location location)
     {
-        cache.invalidate(location);
+        log.debug("Invalidating cache for location: %s", location);
+        cache.asMap().keySet().removeIf(key -> key.location().equals(location));
```

Comment:
I think this will need to change into `CacheUtils.invalidateAllIf(cache, key -> key.location().equals(location))` to avoid the `UnsupportedOperationException` issue you're seeing. Also, if we're invalidating purely on the basis of location (which might be valid, or maybe not) then maybe we actually want something more similar to `.startsWith` instead of `.equals` on that check?

Reply:
looking at the callers of this method, I see that the schema table name is always available to pass, want me to try and give it a shot to change the signature to `public void invalidate(Location location, SchemaTableName schemaTableName)`?

Reply:
Yeah it looks like location based invalidation is always in the context of some table or partition, so it looks like we can either:
- change the signature to `invalidate(SchemaTableName schemaTableName, Location)` and update all usage sites to pass it
- keep separate `invalidate(Table)` and `invalidatePartition(Partition)` methods and do that unwrapping internally

Similar changes will probably need to be made to `TransactionScopeCachingDirectoryLister` whichever route makes the most sense.

Reply:
There was also a TODO in the `TransactionScopeCachingDirectoryLister` so I am going to go with the first option and change the signature, good to make the cache efficient.  I just pushed up the changes, there is one test failing locally which I am looking at, want to run the rest of the build in GHA though to make sure I didn't miss anything.  Thanks for the review, really appreciate it.